[CORE-12460] - Add mirror topic state to cluster link metadata

michael-redpanda opened the pull request:
Adds mirror topic state to the cluster link metadata table.  Also adds commands to add mirror topics to links and to update their states.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
cl/model: Added mirror topic state

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/cl: Added topic state to table

Added tracking topics to the table.  Added a reverse topic -> id lookup
that is used to find the ID of the link responsible for mirroring a
topic as well as accessors to get the topic state directly from the
topic name.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c/cl: Updated errors

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added add and update mirror topic cmds

Added two new commands to handle creation of mirror topics and updating
the mirror topic state.  Though they are currently the same values, the
creation command may be updated to also add in fields such as selecting
the name of the destination topic if it differs from the source topic
name.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster: Added add and update mirror topic cmds

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/cl: Handling new mirror topic commands

Added support for handling the add mirror topic command and the update
mirror topic state command.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c: Added commands to add update mirror topics

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c: Added rpc handling for add/update mirror topics

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c/cl: Handling add/update mirror topics to fe

Added support in cluster link frontend to handle adding and updating
mirror topics.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
c: Wire in add/update mirror topic cmds

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -224,7 +292,8 @@ ss::future<mutation_result> frontend::do_local_mutation(
 cluster::errc frontend::validate_mutation(const cluster_link_cmd& cmd) const {
-    // Initially for DR, we will only support a single cluster link at a time.
+    // Initially for DR, we will only support a single cluster link at a
+    // time.ybri
```
That `ybri` was very important

piyushredpanda replied:
Haha

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--c00df913-db57-453b-8ea0-62b1884c0e2b-->
<!--TEST RESULTS-->
<!--7b2a2b37-ff6a-441e-bc1f-1b8e8d3726ee-->
<!--TEST RESULTS-->
<!--0dbb6434-5334-4f8c-86bb-d7ed96417622-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67684">build#67684</a></summary>


| test_class                 | test_method                              | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:---------------------------|:-----------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest   | test_node_operations                     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67684#01978ff8-49f2-4661-acd4-3896021bff39 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true}                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67684#01978ff8-2b41-42f6-b635-15f2df869b48 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67917">build#67917</a></summary>


| test_class                    | test_method                               | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------------|:------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest                  | test_connection_flicker                   | {"cloud_storage_type": 1}                                                                                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67917#0197a8b5-c160-49ae-8c52-6612a6fb4af6 | FLAKY         | 20/21    | upstream reliability is '99.66555183946488'. current run reliability is '95.23809523809523'. drift is 4.42746 and the allowed drift is set to 50. The test should PASS |
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67917#0197a8b6-eaf5-4e10-970c-aab891c76218 | FLAKY         | 14/21    | upstream reliability is '70.02427184466019'. current run reliability is '66.66666666666666'. drift is 3.35761 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest      | test_node_operations                      | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67917#0197a8b6-eaf7-4370-bde3-6aae10b47169 | FLAKY         | 20/21    | upstream reliability is '98.56459330143541'. current run reliability is '95.23809523809523'. drift is 3.3265 and the allowed drift is set to 50. The test should PASS  |
| SimpleEndToEndTest            | test_relaxed_acks                         | {"write_caching": false}                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67917#0197a8b6-eaf5-4e10-970c-aab891c76218 | FLAKY         | 20/21    | upstream reliability is '99.00497512437812'. current run reliability is '95.23809523809523'. drift is 3.76688 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68177">build#68177</a></summary>


| test_class                  | test_method                              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:-----------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest | test_consumer_offsets_partition_recovery |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68177#0197c136-da37-4db5-bc05-624746ff1ae2 | FLAKY         | 19/21    | upstream reliability is '98.77750611246944'. current run reliability is '90.47619047619048'. drift is 8.30132 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj commented on src/v/cluster/cluster_link/table.h:
```diff
@@ -78,14 +85,26 @@ class table : public ss::peering_sharded_service<table> {
       flat_hash_map<::cluster_link::model::name_t, ::cluster_link::model::id_t>;
 
+    using topic_name_index_t
+      = absl::flat_hash_map<model::topic, ::cluster_link::model::id_t>;
```
here and everywhere chunked hash map please, so we don't get oversized allocs when someone uses 40k+ topics

rockwotj commented on src/v/cluster/cluster_link/table.h:
```diff
@@ -78,14 +85,26 @@ class table : public ss::peering_sharded_service<table> {
     using name_index_t = absl::
       flat_hash_map<::cluster_link::model::name_t, ::cluster_link::model::id_t>;
 
+    using topic_name_index_t
```
instead of being key'd by string name should this be the topic initial revision? I am not really sure how creating topics will work and then if racing with the topic creation is a situation we care about...

mmaslankaprv replied:
Agree, can we use topic-id in here ? Revision is probably not feasible as we can not extract it from remote cluster and it is Redpanda specific

michael-redpanda replied:
This is why I preferred keeping mirror state in the topic....

michael-redpanda replied:
The usage model in my head was that the topic name woudl be added to this table first and then the topic created.  I guess I can flip it here and have the topic created first and then add it to the table and hope adding it to the table doesn't fail.

michael-redpanda replied:
> Agree, can we use topic-id in here ? Revision is probably not feasible as we can not extract it from remote cluster and it is Redpanda specific

The name is the local (destination) cluster's topic name

michael-redpanda replied:
Talked some with @rockwotj - what we'll do to get around some of this is to:

1. Generate the topic ID and push that into the topic when it's created by cluster linking
2. Keep track of the topic ID in the cluster link table (as well as the mirror state)

This will ensure that we know that the cluster link created that topic and 'owns' it.

rockwotj replied:
Are we doing this in this PR or a followup?

michael-redpanda replied:
We are tracking the destination topic's topic id in this PR - pushing topic ID into topic creation will happen later (I think I'm answering your question?)

rockwotj commented on src/v/cluster_link/model/types.h:
```diff
@@ -90,13 +127,59 @@ struct metadata
+    auto serde_fields() { return std::tie(topic, state); }
+};
+
+struct update_mirror_topic_state_cmd
```
for these new commands, care to add a high level description of them Since this PR is mostly boilerplate, it is not clear how these will be used and who will use them (is this internal or a proxy for a user operation?)

mmaslankaprv commented on src/v/cluster_link/model/types.h:
```diff
@@ -78,6 +106,15 @@ struct connection_config
+struct link_state
+  : serde::envelope<link_state, serde::version<0>, serde::compat_version<0>> {
+    /// The set of topics that are being mirrored by this link and their state
+    absl::flat_hash_map<::model::topic, mirror_topic_state> mirror_topics;
```
use `chunked_hash_map`

mmaslankaprv replied:
consider making `mirro_topic_state` a `struct` it will give us more flexibility in future

rockwotj replied:
+1 to both of these

mmaslankaprv commented on src/v/cluster/errc.h:
```diff
@@ -104,6 +104,10 @@ enum class errc : int16_t {
     cluster_link_invalid_update,
     cluster_link_limit_exceeded,
     cluster_link_service_error,
+    cluster_link_topic_already_being_mirrored,
+    cluster_link_topic_being_mirrored_by_other_link,
+    cluster_link_topic_not_being_mirrored,
+    cluster_link_mirror_topic_name_invalid,
```
can we move those to link specific error category ?

michael-redpanda replied:
~~We need to talk about this, you brought this up before but I don't think it's doable~~

edit:

I think I can get something here...

mmaslankaprv commented on src/v/cluster/cluster_link/table.h:
```diff
@@ -84,8 +91,12 @@ class table : public ss::peering_sharded_service<table> {
     using name_index_t = absl::
       flat_hash_map<::cluster_link::model::name_t, ::cluster_link::model::id_t>;
 
+    using topic_name_index_t
```
I think this shouldn't be a global index, it should be a part `:cluster_link::model::metadata`, otherwise two topics from different source clusters would clash

mmaslankaprv replied:
actually i guess the global index makes sense, we can not have two topics with the same name

mmaslankaprv commented on src/v/cluster_link/model/types.h:
```diff
@@ -136,6 +136,41 @@ struct metadata
     }
 };
+
+struct add_mirror_topic_cmd
```
can we merge the two commands into `upsert_mirror_topic_cmd` ?

mmaslankaprv commented on src/v/cluster/cluster_link/table.cc:
```diff
@@ -300,4 +312,68 @@ cluster::errc table::remove_link(const name_t& name) {
+          id,
+          link_id.value());
+        return cluster::errc::cluster_link_topic_being_mirrored_by_other_link;
+    } else if (link_id && link_id.value() == id) {
```
nit: no need for else

pgellert commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -293,13 +363,15 @@ frontend::validator::validate_mutation(const cluster_link_cmd& cmd) const {
-                "Attempting to create a panda link when the maximum number of "
+                "Attempting to create a panda link when the maximum number "
+                "of "
                 "links ({}) is already reached",
```
nit: the formatting is a bit off here

pgellert commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -313,6 +385,69 @@ frontend::validator::validate_mutation(const cluster_link_cmd& cmd) const {
+                    "'{}'",
+                    cmd.value.topic,
+                    meta->get().name);
+                  return cluster::errc::
+                    cluster_link_topic_already_being_mirrored;
```
minor: Do the callers need to be able to differentiate between `errc::cluster_link_topic_already_being_mirrored` vs `errc::success` (idempotent success)?

michael-redpanda replied:
Fair question... it may be helpful to the operator to know that (rather than just not doing anything silently)

pgellert commented on src/v/cluster/cluster_link/tests/frontend_validation_test.cc:
```diff
@@ -65,6 +69,31 @@ class frontend_validation_test : public seastar_test {
+        cluster::cluster_link_add_mirror_topic_cmd add_cmd{id, std::move(cmd)};
+        auto ec = _validator->validate_mutation(add_cmd);
+        if (ec == cluster::errc::success) {
+            co_await _table.local().apply_update(
+              testing::create_add_mirror_topic_command(
+                add_cmd.key, std::move(add_cmd.value)));
```
Do we need to surface the `std::error_code` returned by `apply_update` here?

michael-redpanda replied:
yeah or at least assert on it being successful

pgellert commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -90,6 +95,28 @@ ss::future<mutation_result> frontend::remove_cluster_link(
 
+ss::future<mutation_result> frontend::add_mirror_topic(
+  id_t id, add_mirror_topic_cmd cmd, model::timeout_clock::time_point timeout) {
+    if (!cluster_link_active(false)) {
```
Should we check the license here? Is the idea that already existing DR links will always be allowed to be kept up to date, only new links cannot be created?

michael-redpanda replied:
Haven't had a decision on what the license sanctions will be yet

pgellert commented on src/v/cluster/cluster_link/table.h:
```diff
@@ -52,6 +52,13 @@ class table : public ss::peering_sharded_service<table> {
     /// Finds link ID by name
     std::optional<::cluster_link::model::id_t>
     find_id_by_name(const ::cluster_link::model::name_t& name) const;
+    /// Finds a link ID by the mirror topic name
+    std::optional<::cluster_link::model::id_t>
+    find_id_by_topic(model::topic_view tp) const;
```
Should we think about this as the name of the "remote cluster's topic" or the "local cluster's topic"? For DR they should be the same, but (1) in the future there might be an ask for topic renaming and (2) maybe we could be more precise (eg. include the topic revision) if we treat this as the "local cluster's topic", since we can use the internal APIs to gather data about the topic instead of being limited by the kafka API.

rockwotj replied:
[YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) on topic renaming :)

The thinking from a conversation I had with yesterday is that we're going to need some form of topic IDs to make all this robust anyways.

michael-redpanda replied:
Well there is prior art to having a prefix added to destination topics (but won't be for this release)

michael-redpanda replied:
Anyways to close this out - we're going to reference everything interlaly by the destination cluster topic's name.  Whether prefixing will be there or not really doesn't impact I think what's in the table.

michael-redpanda committed:
cluster: Created cluster::cluster_link::errc

Massive refactor to use new cluster::cluster_link::errc enum

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Addressed some (not all) comments on the PR - major change is the introduction of `cluster::cluster_link::errc` (@mmaslankaprv )

dotnwat commented on src/v/cluster/cluster_link/table.cc:
```diff
@@ -172,7 +240,22 @@ void table::run_callbacks(id_t id) {
+              link_id.value());
+            return errc::topic_being_mirrored_by_other_link;
+        }
+        _topic_name_index.emplace(t.first, id);
```
seems like you might be able to avoid the double container lookup with something like
```
if (!link_id)
    emplace
elif link_id.value != id
    return error
```
but that's based on the expected behavior here. asserting on the return value of emplace is useful to document expected behavior when the container may already contain the key.

rockwotj commented on src/v/cluster/cluster_link/errc.cc:
```diff
@@ -0,0 +1,99 @@
+}
+} // namespace cluster::cluster_link
+
+auto fmt::formatter<cluster::cluster_link::errc>::format(
+  cluster::cluster_link::errc e, fmt::format_context& ctx) const
```
FWIW for enums there is a simpler `format_as` alternative: https://fmt.dev/11.0/api/#udt

In the version we use it only works for `enum` types, but in a future version we can do it for all types. I only mention that to save some of the boilerplate here.

michael-redpanda replied:
oh that's nice.....  why didn't you tell me that earlier (j/k)

michael-redpanda commented:
Force push:

* Replaced the use of absl::flat_hash_map with chunked_hash_map
  * Required adding in explicit copy methods to models
* Moved `paused` into `link_state`
* Created `mirror_Topic_metadata` - now contains metadata info about a mirror topic including its state

michael-redpanda commented:
Force push:

* Rebase

michael-redpanda committed:
c/cl: Converted flat_hash_map to chunked_hash_map

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Changes to `ss::visit` was causing a build break - fixed that
* Fixed some git commit history things

rockwotj commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -320,6 +387,66 @@ errc frontend::validator::validate_mutation(const cluster_link_cmd& cmd) const {
           }
           return errc::success;
+      },
+      [this](const cluster::cluster_link_add_mirror_topic_cmd& cmd) {
```
Just as a note for other reviewers to understand here:
This validation only works because we take a lock when applying local mutations, otherwise it's possible two commands could race.

rockwotj reviewed (commented):
Mostly LGTM, just a question if we want to merge this as is or if the topic ID stuff comes later?

mmaslankaprv commented on src/v/cluster/cluster_link/table.cc:
```diff
@@ -214,6 +288,18 @@ cluster::cluster_link::errc table::remove_link(const name_t& name) {
+
+    // absl::flat_hash_map does not support std::ranges algorithms directly,
+    // so we manually erase by predicate.
+    chunked_vector<::model::topic> topics_to_remove;
+    for (const auto& it : _topic_name_index) {
+        if (it.second == id) {
+            topics_to_remove.emplace_back(it.first);
+        }
+    }
+    for (const auto& topic : topics_to_remove) {
+        _topic_name_index.erase(topic);
+    }
```
The comment above doesn't hold, we are using chunked hash map, also it should be possible to remove map in one pass. There is an `erase_if` function in `unordered_dense`.

mmaslankaprv replied:
The same patten is used in `upsert_link`

mmaslankaprv commented on src/v/cluster/cluster_link/table.cc:
```diff
@@ -312,4 +324,70 @@ cluster::cluster_link::errc table::remove_link(const name_t& name) {
+    run_callbacks(id);
+    return errc::success;
+
+} // namespace cluster::cluster_link
```
nit: redundant comment

michael-redpanda commented:
Force push:

* Addressed PR comments

pgellert reviewed (approved):
lgtm
